ambry-replication/src/main/java/com/github/ambry/replication/RemoteReplicaInfo.java

```diff
@@ -208,25 +224,45 @@ public boolean equals(Object obj) {
   }
 
   /**
-   * Get the meta data response information received for this replica in the most recent replication cycle.
+   * Get the meta data response received for this replica in the most recent replication cycle.
    * @return exchangeMetadataResponse contains the meta data response (missing keys, token info, local lag from remote, etc.).
    */
   synchronized ReplicaThread.ExchangeMetadataResponse getExchangeMetadataResponse() {
     return exchangeMetadataResponse;
   }
 
   /**
-   * Set the meta data exchange information received for this replica in the most recent replication cycle.
-   * Replica threads calls this method to store the metadata responses during replication cycles.
+   * Set the meta data response received for this replica in the most recent replication cycle.
    * @param exchangeMetadataResponse contains meta data response (missing keys, token info, local lag from remote, etc.).
    */
   synchronized void setExchangeMetadataResponse(ReplicaThread.ExchangeMetadataResponse exchangeMetadataResponse) {
-    // We are having this thread safe to avoid conflict between replica thread setting new exchangeMetadataResponse
-    // and replica threads updating the missing store messages in current exchangeMetadataResponse after they are
-    // written to local store via intra-dc replication (method will be added in future PR).
+    // Synchronized to avoid conflict between replica threads setting new exchangeMetadataResponse received for this replica
+    // and replica threads going through existing metadata response (via updateMissingMessagesInMetadataResponse()) to
+    // to compare newly written messages to store with missing message set in metadata response.
     this.exchangeMetadataResponse = exchangeMetadataResponse;
   }
 
+  /**
+   * Update missing store messages found for this replica in its recent exchange metadata response by comparing
+   * (based on the store key) with messages that are written to store by other replica threads.
+   * @param messagesWrittenToStore list of messages written to local store
+   */
+  synchronized void updateMissingMessagesInMetadataResponse(List<MessageInfo> messagesWrittenToStore) {
+
+    Set<MessageInfo> missingStoreMessages = exchangeMetadataResponse.getMissingStoreMessages();
+    if (missingStoreMessages != null && !missingStoreMessages.isEmpty()) {
+      Set<StoreKey> keysWrittenToStore =
+          messagesWrittenToStore.stream().map(MessageInfo::getStoreKey).collect(Collectors.toSet());
+      Set<MessageInfo> missingMessagesFoundInStore = new HashSet<>();
```

Comment:
Minor: can use  stream().filter().collect()

Reply:
Sure, updated code to use stream().filter().collect()